packages/worker/src/worker-options.ts

```diff
@@ -411,13 +416,100 @@ export function addDefaultWorkerOptions(options: WorkerOptions): WorkerOptionsWi
+ */
+export function inspectSystemResources(): { memory: number; swap: number; heap: number } {
+  const resources = {
+    memory: os.totalmem(),
+    swap: 0,
+    heap: v8.getHeapStatistics().heap_size_limit,
+  };
+
+  if (process.platform === 'linux') {
+    // Check for v2 style cgroups
+    if (fs.existsSync('/sys/fs/cgroup/cgroup.controllers')) {
+      // Maximum size (in bytes) of memory usage allowed by this container. 'max' if no constraint.
+      // Examples:
+      //  - 536870912 => 512 Mb
+      //  - max => no constraint
+      const memoryMax = Number(tryReadFileSync('/sys/fs/cgroup/memory.max'));
+      if (!isNaN(memoryMax)) resources.memory = Math.min(memoryMax, resources.memory);
+
+      // Maximum size (in bytes) of swap usage allowed by this container. The reported size EXCLUDES the size of memory itself.
+      // For example, assuming `docker run --memory 512mb --memory-swap 768mb`, memory.swap.max will report 256mb
+      // Examples:
+      //  - 268435456 => 256 Mb
+      //  - max => swap allowed but unconstrained
+      const swapMax = tryReadFileSync('/sys/fs/cgroup/memory.swap.max');
+      if (swapMax) {
+        if (swapMax === 'max')
+          resources.swap = resources.memory; // Swap is allowed, but no size was specified. Default to memory size
+        else if (!isNaN(Number(swapMax))) resources.swap = Number(swapMax);
+      }
+    }
+
+    // Check for v1 style cgroups: /sys/fs/cgroup/memory/memory.limit_in_bytes
+    if (fs.existsSync('/sys/fs/cgroup/memory/memory.limit_in_bytes')) {
+      // Maximum size (in bytes) of memory usage allowed by this container. '9223372036854771712' if no constraint
+      // Examples:
+      //  - 536870912 => 512 Mb
+      //  - 9223372036854771712 => swap allowed but unconstrained
+
+      const memoryMax = Number(tryReadFileSync('/sys/fs/cgroup/memory/memory.limit_in_bytes'));
+      if (!isNaN(memoryMax)) resources.memory = Math.min(memoryMax, resources.memory);
+
+      // Maximum size (in bytes) of swap usage allowed by this container. The reported size INCLUDES the size of memory itself.
+      // For example, assuming `docker run --memory 512mb --memory-swap 768mb`, memory.memsw.limit_in_bytes will report 768mb.
+      // File might also not exists if system does not support swap.
+      // Defaults to '9223372036854771712' if no constraint.
+      // Examples:
+      //  - 536870912 => 512 Mb
+      //  - 9223372036854771712 => no constraint
+      const swapMax = Number(tryReadFileSync('/sys/fs/cgroup/memory/memory.memsw.limit_in_bytes'));
+      if (!isNaN(swapMax)) {
+        if (swapMax === 9223372036854771712)
+          resources.swap = resources.memory; // Swap is allowed, but no size was specified. Default to memory size
+        else resources.swap = swapMax - resources.memory; // memory.memsw.limit_in_bytes includes mem+swap
+      }
+    }
+  }
+
+  if (resources.heap > resources.memory * 0.5) {
```

Comment:
Curious why 50%

Reply:
I checked node's official recommandations on this, and they recommand setting `--max-old-space-size` at 75% of total memory, which brings `heap_size_limit` at approximately 80%. I wouldn't mind increasing the warning threashold at 80%. This should be ok in most case, but we risk having containers getting killed from times to times due to OOM. That depends a lot on actual usage patterns.

Reply:
The problem is not what is _inside_ the heap, but rather what is _outside_ of the heap...

Let's break it down... Essentially, we have following equations:

```
container memory usage = node process memory usage + other processes memory usage

node process memory usage = node heap memory + non-heap memory
node heap memory = young spaces + mid spaces + old space + compiled code + large objects + other heap stuff
node non-heap memory = node stacks + memory allocated from Buffer + memory allocated from native code + other non-heap stuff
```

If `container memory usage` exceeds the (hard) memory limit set on the container, the container immediately gets killed by the OS. No warning and no chance to run GC.

So the question here is essentially: how much space do we reasonably need to set asside for all non-heap to be _safe_.